A33-Fault-Injection.md

```diff
@@ -0,0 +1,253 @@
+
+**Response rate limit algorithm:** Envoy uses the token bucket algorithm to throttle the speed of letting response pass through this filter. It will try to write as much as possible from the buffer, then it will schedule another write when the next token is available.
+
+Each token permits to write `(max_kbps * 1024) / SecondDivisor` of bytes. The `SecondDivisor` is a constant 16, which splits 1 second into 16 segments (~63ms apart).
+
+Only the body of the HTTP stream is throttled by the response rate limit algorithm. Not headers or trailers.
+
+**Response rate limit and timeout:** Envoy implements fault injection is an `HTTPFilter`, which can choose to operate on bytes or HTTP messages or events. It is unclear if gRPC can perform the precise same behavior, since gRPC generally operates on proto messages (Core has visibility to bytes). The timeout timer may start at the beginning of an RPC, or the receipt of the first bytes from peer, or after the entire message is received.
+
+
+### Absence of Envoy Runtime in gRPC
+
+Envoy has a mechanism to allow reload of configuration without restarting the entire process named [Runtime](https://www.envoyproxy.io/docs/envoy/latest/operations/runtime). It can be loaded from a static file, updated with a local file system, or overridden from an Admin console ([reference](https://www.envoyproxy.io/docs/envoy/latest/configuration/operations/runtime)).
+
+However, as a library, gRPC doesn’t have a decent standing to maintain an “Runtime” state. Available options are 1) ignore Runtime settings; 2) allow initial Runtime setting (from a file or env variables). According to previous design decisions, the Runtime settings are usually ignored (see [gRPC xDS Traffic Splitting and Routing](https://github.com/grpc/proposal/blob/master/A28-xds-traffic-splitting-and-routing.md)).
+
+### Fault Injection HTTP Headers in Envoy
+
+Envoy allows applications to specify [HTTP headers](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/fault_filter#controlling-fault-injection-via-http-headers) to enable fault injection. This feature still needs a fault injection filter to be specified, and the delay or the abort field needs to explicitly say `HeaderDelay` or `HeaderAbort`.
+
+### Prior Art: Envoy HTTP Filter
+
+Envoy filters have very granular logic injection points at different stages of a request (or stream). For example, while decoding headers / data / trailer, encoding headers / data / trailer, or completion of decoding / encoding. In each injection point, the return value dictates whether Envoy should proceed to other filters or stop. This design gives immense flexibility for filters. The delay injection could be a headache since it can’t block the thread. But fault injection filters utilize this design and are able to schedule callback (rest of request processing logic) asynchronous.
+
+## Proposal
```

Comment:
I think this section should start out with a high-level overview of what this gRFC is proposing.  This should be similar to the material that you currently have in the "background" section above, but much more concise.  Specifically, I think we should briefly touch on the following high-level points:

- We will support delays before starting a request, failing with a particular gRPC status code, and imposing a limit on the maximum number of currently active faults.
- We will allow these features to be configured directly from the xDS config.  We will also allow the xDS config to tell us to look at request headers for configuration.
- We will not support configuration based on global runtime settings, because unlike Envoy, gRPC does not have the concept of a global runtime.
- We will not initially support response rate limiting, since that will require more design and implementation work.  We will wait for a use-case to present itself before we do that work.
- We will not support selecting faults based on upstream or downstream peers until a use-case presents itself.